src/MuTalk-Model/MTSubclassReplacementOperator.class.st

```diff
@@ -0,0 +1,93 @@
+Class {
+	#name : 'MTSubclassReplacementOperator',
+	#superclass : 'MTBlockBasedMutantOperator',
+	#instVars : [
+		'classesToReplace'
+	],
+	#category : 'MuTalk-Model-Operators',
+	#package : 'MuTalk-Model',
+	#tag : 'Operators'
+}
+
+{ #category : 'printing' }
+MTSubclassReplacementOperator >> description [
+
+	^ 'Replace a class reference by a reference to one of its subclasses'
+]
+
+{ #category : 'applying' }
+MTSubclassReplacementOperator >> expressionToReplace [
+
+	^ [ :aNode |
+	  aNode isVariable and: [
+		  (self class environment includesKey: aNode name) and: [
+			  | referencedClass shouldReplace |
+			  referencedClass := self class environment at: aNode name.
+			  shouldReplace := referencedClass isClass and: [
+				                   referencedClass subclasses notEmpty ].
+			  shouldReplace ifTrue: [ classesToReplace add: referencedClass ].
+			  shouldReplace ] ] ]
+]
+
+{ #category : 'initialization' }
+MTSubclassReplacementOperator >> initialize [
+
+	super initialize.
+	classesToReplace := OrderedCollection new
+]
+
+{ #category : 'private' }
+MTSubclassReplacementOperator >> modifiedSourceFor: aCompiledMethod number: aNumber newExpression: anExpression [
+
+	^ self
+		  modifiedSourceFor: aCompiledMethod
+		  with: aCompiledMethod parseTree
+		  number: aNumber
+		  newExpression: anExpression
+]
+
+{ #category : 'private' }
+MTSubclassReplacementOperator >> modifiedSourceFor: aCompiledMethod with: aParseTree number: aNumber newExpression: anExpression [
+
+	| parser allNodes |
+	parser := aParseTree copy.
+	allNodes := parser allChildren select: self expressionToReplace.
+	allNodes size >= aNumber ifTrue: [
+		| oldNode newNode |
+		oldNode := allNodes at: aNumber.
+		newNode := anExpression value: oldNode.
+		oldNode replaceWith: newNode ].
+	^ parser formattedCode
```

Comment:
Hmm is this really different from the superclass method?

Reply:
Not so much indeed, I will refactor a bit